src/php/NodeEnvironment.php

```diff
@@ -179,4 +184,16 @@ public function getBoundTo(): string
     {
         return $this->boundTo;
     }
+
+    public function isDefAllowed(): bool
+    {
+        return $this->defAllowed;
+    }
+
+    public function setDefAllowed(bool $defAllowed): self
```

Comment:
NodeEnvironment should always be immutable. So instead of a setter the method should be called `withDefAllowed` and return a new instance of NodeEnvironment (see other methods in this class).

I made it immutable because it is much easier to reason about it.

Reply:
Absolutely right, good catch! I will change it! :)